File: python/mirage/commands/builtin/gdrive/tail/tail.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import AsyncIterator

from mirage.accessor.gdrive import GDriveAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.gdrive._provision import file_read_provision
from mirage.commands.builtin.tail_helper import _parse_n, tail_bytes
from mirage.commands.builtin.utils.stream import _read_stdin_async
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.gdrive.glob import resolve_glob
from mirage.core.gdrive.read import read as gdrive_read
from mirage.io.types import ByteSource, IOResult
from mirage.provision.types import ProvisionResult
from mirage.types import PathSpec


async def tail_provision(
    accessor: GDriveAccessor,
    paths: list[PathSpec],
    *texts: str,
    index: IndexCacheStore = None,
    **_extra: object,
) -> ProvisionResult:
    return await file_read_provision(
        accessor,
        paths,
        "tail " + " ".join(p.original if isinstance(p, PathSpec) else p
                           for p in paths),
        index=index)


async def _tail_result(raw: bytes, lines: int, plus_mode: bool,
                       bytes_mode: int | None) -> AsyncIterator[bytes]:
    if bytes_mode is not None:
        yield raw[-bytes_mode:] if bytes_mode else b""
        return
    yield tail_bytes(raw, lines, plus_mode=plus_mode)


@command("tail",
         resource="gdrive",
         spec=SPECS["tail"],
         provision=tail_provision)
async def tail(
    accessor: GDriveAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    n: str | None = None,
    c: str | None = None,
    q: bool = False,
    v: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    lines, plus_mode = _parse_n(n)
    bytes_mode = int(c) if c is not None else None
    if paths:
        paths = await resolve_glob(accessor, paths, index)
        p = paths[0]
        raw = await gdrive_read(accessor, p, index)
        return _tail_result(raw, lines, plus_mode, bytes_mode), IOResult()
    raw = await _read_stdin_async(stdin)
    if raw is None:
        raise ValueError("tail: missing operand")
    return _tail_result(raw, lines, plus_mode, bytes_mode), IOResult()

File: typescript/packages/core/src/commands/builtin/mongodb/tail.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { MongoDBAccessor } from '../../../accessor/mongodb.ts'
import type { IndexCacheStore } from '../../../cache/index/store.ts'
import { findDocuments } from '../../../core/mongodb/_client.ts'
import { resolveGlob } from '../../../core/mongodb/glob.ts'
import { streamAny } from '../../../core/mongodb/read.ts'
import { detectScope } from '../../../core/mongodb/scope.ts'
import { applyElision, elisionPaths, stringifyDoc } from '../../../core/mongodb/stream.ts'
import { ScopeLevel } from '../../../core/mongodb/types.ts'
import { type PathSpec, ResourceName } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { tailGeneric } from '../generic/tail.ts'
import { parseN } from '../tail_helper.ts'
import { fileReadProvision } from './_provision.ts'

const ENC = new TextEncoder()

// Fetches only the last N documents server-side (sort _id desc + limit)
// instead of streaming the whole collection; tailGeneric then trims the
// already-small chunk. Falls back to a full stream for byte mode, +N mode,
// and non-collection paths.
async function* tailSource(
  accessor: MongoDBAccessor,
  p: PathSpec,
  index: IndexCacheStore | undefined,
  lines: number,
  pushdown: boolean,
): AsyncIterable<Uint8Array> {
  const scope = detectScope(p)
  if (
    pushdown &&
    scope.level === ScopeLevel.DOCUMENTS &&
    scope.database !== null &&
    scope.name !== null
  ) {
    const limit = Math.min(lines, accessor.config.maxDocLimit)
    const docs = await findDocuments(
      accessor,
      scope.database,
      scope.name,
      {},
      { limit, sort: { _id: -1 } },
    )
    docs.reverse()
    if (docs.length === 0) return
    const elide = elisionPaths(accessor, scope.database, scope.name)
    const jsonl =
      docs.map((d) => stringifyDoc(elide.size > 0 ? applyElision(d, elide) : d)).join('\n') + '\n'
    yield ENC.encode(jsonl)
    return
  }
  yield* streamAny(accessor, p, index)
}

async function tailCommand(
  accessor: MongoDBAccessor,
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const resolved =
    paths.length > 0 ? await resolveGlob(accessor, paths, opts.index ?? undefined) : []
  const nRaw = typeof opts.flags.n === 'string' ? opts.flags.n : null
  const [lines, plusMode] = parseN(nRaw)
  const pushdown = typeof opts.flags.c !== 'string' && !plusMode && lines > 0
  return tailGeneric(resolved, texts, opts, (p) =>
    tailSource(accessor, p, opts.index ?? undefined, lines, pushdown),
  )
}

export const MONGODB_TAIL = command({
  name: 'tail',
  resource: ResourceName.MONGODB,
  spec: specOf('tail'),
  fn: tailCommand,
  provision: fileReadProvision,
})
